### PR TITLE
Add support for GCP workload identity

### PR DESCRIPTION
**Pull request by fiunchinho**
Towards https://github.com/giantswarm/roadmap/issues/1313

The same way we have support for IRSA when authenticating with AWS, we want to use workload identity when running on GCP clusters.

---

## Checklist

- [X] Added a CHANGELOG entry

### CODE REVIEW

File: helm/external-dns-app/templates/serviceaccount.yaml [hunk 1]
```diff
@@ -6,6 +6,9 @@ metadata:
   {{- if and (eq .Values.provider "aws") (eq .Values.aws.irsa "true") (eq .Values.aws.access "internal") }}
   annotations:
     eks.amazonaws.com/role-arn: "arn:aws:iam::{{ .Values.aws.accountID }}:role/{{ template "aws.iam.role" . }}"
+  {{- else if and (eq .Values.provider "gcp") }}
```
Comment (webwurst) [inline helm/external-dns-app/templates/serviceaccount.yaml:9]: Maybe you also want to check here if `.Values.gcpProject` is set?

Comment (fiunchinho) [reply helm/external-dns-app/templates/serviceaccount.yaml]: Yeah. Take a look now.

**Review by webwurst — APPROVED** [review]
Thanks!

File: helm/external-dns-app/templates/deployment.yaml [hunk 2]
```diff
@@ -11,12 +11,18 @@ spec:
   selector:
     matchLabels:
       {{- include "labels.selector" . | nindent 6 }}
+      {{- if and (eq .Values.provider "gcp") (.Values.gcpServiceAccountID) (.Values.gcpWorkloadIdentityPool) (.Values.gcpIdentityProvider) }}
+      giantswarm.io/gcp-workload-identity: enabled
```
Comment (Gacko) [inline helm/external-dns-app/templates/deployment.yaml:15]: This will probably break the rollout on existing clusters - why is this required anyway?

File: helm/external-dns-app/templates/deployment.yaml [hunk 3]
```diff
@@ -11,12 +11,18 @@ spec:
...
+      {{- end }}
   strategy:
     type: Recreate
   template:
     metadata:
       labels:
         {{- include "labels.common" . | nindent 8 }}
+        {{- if and (eq .Values.provider "gcp") (.Values.gcpServiceAccountID) (.Values.gcpWorkloadIdentityPool) (.Values.gcpIdentityProvider) }}
+        giantswarm.io/gcp-workload-identity: enabled
```
Comment (Gacko) [inline helm/external-dns-app/templates/deployment.yaml:24]: Rather move this to the `labels.common` template.

**Review by Gacko — CHANGES_REQUESTED** [review]

File: helm/external-dns-app/templates/deployment.yaml (hunk 3 above)
Comment (fiunchinho) [reply helm/external-dns-app/templates/deployment.yaml:24]: I don't want the label on any other resource, only on the pods.

File: helm/external-dns-app/templates/deployment.yaml (hunk 2 above)
Comment (Gacko) [reply helm/external-dns-app/templates/deployment.yaml:15]: Don't get me wrong: For me `giantswarm.io/gcp-workload-identity` rather sounds like a property than a criteria.

Comment (fiunchinho) [reply helm/external-dns-app/templates/deployment.yaml:15]: There is an admission controller that is watching `Pods` to inject the credentials for workload identity. The admission controller only receive requests for `Pods` with this label, because annotations can't be used to select resources. So we need the label rather than an annotation.

Comment (ubergesundheit) [reply helm/external-dns-app/templates/deployment.yaml:15]: I think @Gacko is right. Don't change the `matchLabels`. The change below should be enough

Comment (Gacko) [reply helm/external-dns-app/templates/deployment.yaml:15]: You can still add this label to the normal labels section but please do not add it to the selector labels. Labels like `app.kubernetes.io/name`, `app.kubernetes.io/instance` and `app.kubernetes.io/component` normally don't change because an app's name, instance or component is normally not bound to any kind of function nor describing how the app works and therefore can be used as selector labels quite well.

`giantswarm.io/gcp-workload-identity` to me sounds like a label which might change in the future or at least does not differentiate several deployments of this app. So as long as there won't be several deployments of this app with several unique identifying values other than `enabled` and required for telling one pod apart from another around, I definitely wouldn't make it a selector label.

Comment (fiunchinho) [reply helm/external-dns-app/templates/deployment.yaml]: Ok, I removed it. Thanks a lot for the feedback.

**Review by Gacko — APPROVED** [review]

**Comment by paurosello** [comment]
@fiunchinho anything against merging this?

File: helm/external-dns-app/values.schema.json [hunk 4]
```diff
@@ -12,19 +12,19 @@
                     "type": "string"
                 },
                 "baseDomain": {
-                    "type": ["null", "string"]
+                    "type": "null"
```
Comment (mcharriere) [inline helm/external-dns-app/values.schema.json:15]: careful with these (https://intranet.giantswarm.io/docs/organizational-structure/teams/cabbage/app-updates/helm-values-schema/#caveats)

**Review by mcharriere — CHANGES_REQUESTED** [review]

File: helm/external-dns-app/values.schema.json (hunk 4 above)
Comment (mnitchev) [reply helm/external-dns-app/values.schema.json]: I've fixed all of them now. Please re-review

**Review by mcharriere — APPROVED** [review]
lgtm!
